File: April2022/Java/3SumWithMultiplicity.java

// @saorav21994
// TC : O(max value ^ 2)
// SC : O(max value)

class Solution {
    public int threeSumMulti(int[] arr, int target) {
        long [] count = new long[101];
        long res = 0L;
        int mod = 1000000007;
        for (int i : arr) 
            count[i]++;
        for (int i = 0; i <= 100; i++) {
            for (int j = i; j <= 100; j++) {
                int k = target-i-j;
                if (k > 100 || k < 0) 
                    continue;
                else if (i == j && j == k) {
                    res += count[i] * (count[i]-1) * (count[i]-2) / 6;
                    res %= mod;
                }
                else if (i == j && j != k) {
                    res += count[i] * (count[i]-1) / 2 * count[k];
                    res %= mod;
                }
                else if (j < k) {
                    res += count[i] * count[j] * count[k];
                    res %= mod;
                }
            }
        }
        return (int)(res%mod);
    }
}

File: April2022/Kotlin/3SumwithMultiplicity.kt
// Author: Shobhit Behl (LC: shobhitbruh)
class Solution {
    fun threeSumMulti(arr: IntArray, target: Int): Int {
        var hm=HashMap<Int,Long>();
        for(el in arr){
            hm.put(el,hm.getOrDefault(el,0L)+1L);
        }
        var ans:Long=0;
        for(e1 in hm.keys){
            for(e2 in hm.keys){
                var e3:Int=target-(e1+e2);
                if(hm.containsKey(e3)){
                     val f1 = hm[e1]!!
                    val f2 = hm[e2]!!
                    val f3 = hm[e3]!!
                    if (e1 == e2 && e2 == e3) {
                        ans += f1 * (f1 - 1) * (f1 - 2) / 6
                    } else if (e1 == e2 && e2 != e3) {
                        ans += f1 * (f1 - 1) / 2 * f3
                    } else if (e1 < e2 && e2 < e3) {
                        ans += f1 * f2 * f3
                    }
                }
                ans=ans%1000000007;
            }
        }
        return ans.toInt();
    }
}